### PR TITLE
[DM-23942] Rewrite using aiohttp and aioredis

### PR DESCRIPTION
**Pull request by rra**
- Convert to aiohttp
- Test on Python 3.8 as well as 3.7
- Store sessions in an encrypted cookie that is reset on application restart (for now)
- Replace Flask flash messages with something more manual for now
- Use aiohttp to retrieve the JWKS for an issuer
- Convert to aioredis

### CODE REVIEW

File: src/jwt_authorizer/verify.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+            url = urljoin(issuer.url, ".well-known/jwks.json")
+
+        r = await self.get_url(url)
+        if r.status != 200:
+            msg = f"Cannot retrieve keys from {issuer.url}"
+            raise KeyClientException(msg)
+
+        body = await r.json()
+        if "keys" not in body:
+            msg = f"No keys property in JWKS metadata for {issuer.url}"
```
Comment (cbanek) [inline src/jwt_authorizer/verify.py:91]: Small nitpick, but it looks like you're discovering the url then defaulting to the issuer url, but you don't say what URL you're getting in the errors (you use issuer.url), might be helpful to emit the URL somewhere, but it will probably never happen.

File: src/jwt_authorizer/verify.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+        return public_key
+
+    @staticmethod
+    def _build_public_key(exponent: int, modulus: int) -> bytes:
+        """Convert an exponent and modulus to a PEM-encoded key."""
+        components = rsa.RSAPublicNumbers(exponent, modulus)
+        public_key = components.public_key(backend=default_backend())
+        return public_key.public_bytes(
+            encoding=Encoding.PEM, format=PublicFormat.SubjectPublicKeyInfo,
+        )
```
Comment (cbanek) [inline src/jwt_authorizer/verify.py:282]: fancy, so this is the python version of ssh-keygen -y!

File: src/jwt_authorizer/verify.py [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+    Returns
+    -------
+    token_verifier : `TokenVerifier`
+        A TokenVerifier created from that Flask application configuration.
+    """
+    logger: Logger = request["safir/logger"]
+    config: Config = request.config_dict["jwt_authorizer/config"]
+
+    if "jwt_authorizer/key_client" in request.config_dict:
+        key_client = request.config_dict["jwt_authorizer/key_client"]
```
Comment (cbanek) [inline src/jwt_authorizer/verify.py:302]: is this set somewhere else?  I'm trying to do some searching but I don't see this used elsewhere?

File: tests/handlers/analyze_test.py [hunk 4]
```diff
@@ -7,59 +7,57 @@
...
-    payload = {
-        "aud": "https://test.example.com/",
-        "email": "some-user@example.com",
-        "iss": "https://orig.example.com/",
-        "jti": "some-unique-id",
-        "sub": "some-user",
-        "uidNumber": "1000",
-    }
+async def test_analyze_ticket(aiohttp_client: TestClient) -> None:
     ticket = Ticket()
```
Comment (cbanek) [inline tests/handlers/analyze_test.py:23]: is there a double ticket = Ticket()?

File: tests/handlers/analyze_test.py [hunk 5]
```diff
@@ -7,59 +7,57 @@
...
+        params={"capability": "exec:admin"},
+        headers={"Authorization": f"Bearer {token}"},
+        cookies={"oauth2_proxy": cookie},
+    )
+    assert r.status == 200
+    token = r.headers["X-Auth-Request-Token"]
+
+    # Now pass that token to the /auth/analyze endpoint.
+    r = await client.post(
+        "/auth/analyze", data={"token": ticket.encode("oauth2_proxy")},
```
Comment (cbanek) [inline tests/handlers/analyze_test.py:54]: Is this supposed to be token instead of ticket?  A bit confused by by the comment.

File: tests/handlers/auth_test.py [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+
+    r = await client.get(
+        "/auth",
+        params={"capability": "exec:admin"},
+        headers={"Authorization": "Bearer"},
+    )
+    assert r.status == 401
+    assert r.headers["WWW-Authenticate"]
+    assert_www_authenticate_header_matches(
+        r.headers["WWW-Authenticate"], "Bearer", "Unable to find token"
```
Comment (cbanek) [inline tests/handlers/auth_test.py:54]: is this a different case than passing in an invalid token?

File: tests/handlers/auth_test.py [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+
+    r = await client.get(
+        "/auth",
+        params={"capability": "exec:admin"},
+        headers={"Authorization": f"Bearer {token}"},
+    )
+    assert r.status == 403
+    body = await r.text()
+    assert "No Capability group found in user's `isMemberOf`" in body
+    assert r.headers["X-Auth-Request-Token-Capabilities"] == ""
```
Comment (cbanek) [inline tests/handlers/auth_test.py:83]: ah looks like the headers are created and empty - unless it's maybe different over the wire

File: tests/handlers/auth_test.py [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+    )
+    assert r.status == 403
+    body = await r.text()
+    assert "No Capability group found in user's `isMemberOf`" in body
+    assert r.headers["X-Auth-Request-Token-Capabilities"] == "exec:test"
+    assert (
+        r.headers["X-Auth-Request-Capabilities-Accepted"]
+        == "exec:admin exec:test"
+    )
+    assert r.headers["X-Auth-Request-Capabilities-Satisfy"] == "all"
```
Comment (cbanek) [inline tests/handlers/auth_test.py:107]: This is probably scope creep, but just thinking out loud, it might be useful to try to collect logs of issues with capabilities if we decide to go that way.  That might make it more obvious when your jobs or things are going wrong or lack the proper capabilities when the user might not be directly connected to the thing doing the work and getting the error.  Just a thought.

File: tests/handlers/auth_test.py [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+    token = create_test_token(keypair, ["test"])
+    app = await create_test_app(keypair)
+    client = await aiohttp_client(app)
+
+    r = await client.get(
+        "/auth",
+        params=[
+            ("capability", "exec:admin"),
+            ("capability", "exec:test"),
+            ("satisfy", "any"),
```
Comment (cbanek) [inline tests/handlers/auth_test.py:146]: I'm a bit confused by this test... satisfy any, and then the capabilities on :151 doesn't have the admin in there?  Is that the capability it's picking to satisfy something?

File: tests/handlers/auth_test.py [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+    r = await client.get(
+        "/auth",
+        params={"capability": "exec:admin"},
+        headers={
+            "Authorization": "Basic blah",
+            "X-Forwarded-Access-Token": token,
+        },
+    )
+    assert r.status == 200
+    assert r.headers["X-Auth-Request-Email"] == "some-user@example.com"
```
Comment (cbanek) [inline tests/handlers/auth_test.py:175]: So this test proves that the auth header isn't used by the forwarded token is used?

File: tests/handlers/auth_test.py [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+
+    basic = f"x-oauth-basic:{token}".encode()
+    basic_b64 = base64.b64encode(basic).decode()
+    r = await client.get(
+        "/auth",
+        params={"capability": "exec:admin"},
+        headers={"Authorization": f"Basic {basic_b64}"},
+    )
+    assert r.status == 200
+    assert r.headers["X-Auth-Request-Email"] == "some-user@example.com"
```
Comment (cbanek) [inline tests/handlers/auth_test.py:213]: There's also the token is the username and no password case right?

File: tests/handlers/auth_test.py [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+        "aud": "https://example.com/api",
+        "email": "some-user@example.com",
+        "exp": ANY,
+        "iat": ANY,
+        "isMemberOf": [{"name": "admin"}],
+        "iss": "https://test.example.com/",
+        "jti": ticket.as_handle("oauth2_proxy"),
+        "sub": "some-user",
+        "uid": "some-user",
+        "uidNumber": "1000",
```
Comment (cbanek) [inline tests/handlers/auth_test.py:342]: Since you've got this token here and in the test above and it's got just static data might want to put that in a function as what the default test token representation looks like

File: tests/handlers/tokens_test.py [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+    r = await client.post(
+        f"/auth/tokens/{handle}",
+        headers={"X-Auth-Request-Token": token},
+        data={"method_": "DELETE", "_csrf": csrf_token},
+    )
+    assert r.status == 200
+    body = await r.text()
+    assert f"token with the ticket_id {handle} was deleted" in body
+
+    assert await token_store.get_tokens("1000") == []
```
Comment (cbanek) [inline tests/handlers/tokens_test.py:161]: so this 1000 is the uid right?  Maybe stick some of these magic variables together with your test token parts just to make it even more clear with some of these test data things sprinkled about.

File: tests/util.py [hunk 14]
```diff
@@ -17,14 +22,83 @@
...
+
+    Returns
+    -------
+    result : `bytes`
+        The equivalent URL-safe base64-encoded string corresponding to the
+        number in big endian order.
+    """
+    bit_length = data.bit_length()
+    byte_length = bit_length // 8 + 1
+    data_as_bytes = data.to_bytes(byte_length, byteorder="big", signed=False)
```
Comment (cbanek) [inline tests/util.py:49]: Never thought I'd have to think about endian issues in python, but it's good to know. :)

File: Makefile [hunk 15]
```diff
@@ -2,7 +2,7 @@
 update-deps:
 	pip install --upgrade pip-tools pip setuptools
 	pip-compile --upgrade --build-isolation --generate-hashes --output-file requirements/main.txt requirements/main.in
-	pip-compile --upgrade --build-isolation --generate-hashes --output-file requirements/dev.txt requirements/dev.in
+	pip-compile --upgrade --build-isolation --allow-unsafe --generate-hashes --output-file requirements/dev.txt requirements/dev.in
```
[the remark this answers is not in the document]
Comment (rra) [reply Makefile:5]: It's because of sphinx, which depends on setuptools. I've added a comment to `Makefile` to explain why this is here.

File: requirements/main.in [hunk 16]
```diff
@@ -5,15 +5,21 @@
 # After editing, update requirements/main.txt by running:
 #     make update-deps
 
+aiohttp
+aiodns
+chardet
```
[the remark this answers is not in the document]
Comment (rra) [reply requirements/main.in:10]: Yeah, this is irritating because those dependencies are really part of the aiohttp dependencies, which is why they're grouped. I'll add a comment to make that clearer.

File: src/jwt_authorizer/authnz.py [hunk 17]
```diff
@@ -198,8 +215,8 @@ def verify_authorization_strategy() -> Tuple[List[str], str]:
         must be present.
     """
     # Authorization Checks
-    capabilities = request.args.getlist("capability")
-    satisfy = request.args.get("satisfy") or "all"
+    capabilities = request.query.getall("capability")
+    satisfy = request.query.get("satisfy") or "all"
```
[the remark this answers is not in the document]
Comment (rra) [reply src/jwt_authorizer/authnz.py:219]: This is just a difference between aiohttp and Flask. Flask uses `request.args` and names the functions differently.

File: src/jwt_authorizer/config.py [hunk 18]
```diff
@@ -2,140 +2,300 @@
...
+        group_mapping = {}
+        if settings.get("GROUP_MAPPING"):
+            for key, value in settings["GROUP_MAPPING"].items():
+                assert isinstance(key, str), "group_mapping is malformed"
+                assert isinstance(value, list), "group_mapping is malformed"
+                group_mapping[key] = value
+
+        store_session_settings = settings["OAUTH2_STORE_SESSION"]
+        if store_session_settings.get("OAUTH2_PROXY_SECRET"):
+            secret_b64 = store_session_settings["OAUTH2_PROXY_SECRET"]
```
[the remark this answers is not in the document]
Comment (rra) [reply src/jwt_authorizer/config.py:194]: Yeah, this is just for testing because creating a bunch of files in the test environment is kind of irritating. That's a good point about having one way to do it, though. I'll revisit this when it comes time to rewrite the configuration system to use pydantic, which will mean revisiting all the configuration decisions.

File: src/jwt_authorizer/config.py [hunk 19]
```diff
@@ -2,140 +2,300 @@
...
+        session_store = SessionStoreConfig(
+            ticket_prefix=store_session_settings["TICKET_PREFIX"],
+            redis_url=store_session_settings["REDIS_URL"],
+            oauth2_proxy_secret=secret,
+        )
+
+        if settings.get("KNOWN_CAPABILITIES"):
+            known_capabilities = settings["KNOWN_CAPABILITIES"]
+        else:
+            known_capabilities = {}
```
[the remark this answers is not in the document]
Comment (rra) [reply src/jwt_authorizer/config.py:209]: Good call. Done.

File: src/jwt_authorizer/config.py [hunk 20]
```diff
@@ -2,140 +2,300 @@
...
+
+        issuers = {}
+        if settings.get("ISSUERS"):
+            for url, info in settings["ISSUERS"].items():
+                audience_setting = info["AUDIENCE"]
+                if isinstance(audience_setting, str):
+                    audience: Tuple[str, ...] = (audience_setting,)
+                else:
+                    audience = tuple(audience_setting)
```
[the remark this answers is not in the document]
Comment (rra) [reply src/jwt_authorizer/config.py:218]: It can be either a string or a list. I'll add a comment to make that clearer. (This is something else that I'll fix properly when redoing the configuration code. It's supported just for backward compatibility.)

File: src/jwt_authorizer/config.py [hunk 21]
```diff
@@ -2,140 +2,300 @@
...
+    """Configuration for JWT Authorizer."""
+
+    realm: str
+    """Realm for HTTP authentication."""
+
+    authenticate_type: AuthenticateType
+    """What type of authentication to request in WWW-Authenticate."""
+
+    loglevel: Optional[str]
+    """Log level, chosen from the string levels supported by logging."""
```
[the remark this answers is not in the document]
Comment (rra) [reply src/jwt_authorizer/config.py:126]: Yeah, this is another temporary wart that I need to clean up when the configuration handling is rewritten. Log level in the main configuration isn't used currently. The two configuration structs need to be reunified. This is just a temporary hack.

File: src/jwt_authorizer/handlers/auth.py [hunk 22]
```diff
@@ -0,0 +1,304 @@
...
+    elif "x-forwarded-ticket-id-token" in request.headers:
+        encoded_token = request.headers["x-forwarded-ticket-id-token"]
+    elif auth_type.lower() == "basic":
+        logger.debug("Using OAuth with Basic")
+        # We fallback to user:token. We ignore the user.
+        # The Token is in the password
+        encoded_basic_auth = auth_blob
+        basic_auth = base64.b64decode(encoded_basic_auth)
+        user, password = basic_auth.strip().split(b":")
+        if password == b"x-oauth-basic":
```
[the remark this answers is not in the document]
Comment (rra) [reply src/jwt_authorizer/handlers/auth.py:249]: This was copied from existing code (GitHub doesn't note that very clearly), but good call that it's pretty confusing. I'll break this out into a separate function to decode a Basic Auth string. Down the road, I hope to support fewer different ways of providing the token.

File: src/jwt_authorizer/handlers/auth.py [hunk 23]
```diff
@@ -0,0 +1,304 @@
...
+            headers["X-Auth-Request-Uid"] = uid
+        if groups_list:
+            groups = ",".join([g["name"] for g in groups_list])
+            headers["X-Auth-Request-Groups"] = groups
+
+    encoded_token, oauth2_proxy_ticket = await _check_reissue_token(
+        request, encoded_token, verified_token
+    )
+    headers["X-Auth-Request-Token"] = encoded_token
+    headers["X-Auth-Request-Token-Ticket"] = oauth2_proxy_ticket
```
[the remark this answers is not in the document]
Comment (rra) [reply src/jwt_authorizer/handlers/auth.py:329]: It makes an empty header. This was the existing behavior, so I kept it, although I want to clean this up later when I start getting rid of backward compatibility stuff.

File: src/jwt_authorizer/handlers/tokens.py [hunk 24]
```diff
@@ -0,0 +1,276 @@
...
+)
+
+if TYPE_CHECKING:
+    from jwt_authorizer.config import Config
+    from logging import Logger
+    from typing import Dict
+
+__all__ = [
+    "get_token_by_handle",
+    "get_tokens",
```
[the remark this answers is not in the document]
Comment (rra) [reply src/jwt_authorizer/handlers/tokens.py:32]: That's alphabetical according to sort! `_` sorts before `s`.

File: src/jwt_authorizer/handlers/tokens.py [hunk 25]
```diff
@@ -0,0 +1,276 @@
...
+        decoded_token = await authenticate(request, encoded_token)
+    except PyJWTError as e:
+        logger.exception("Failed to authenticate token")
+        raise unauthorized(request, "Invalid token", str(e))
+
+    capabilities = config.known_capabilities
+    form = api_capabilities_token_form(capabilities, await request.post())
+
+    if not form.validate():
+        return {"form": form, "capabilities": capabilities}
```
[the remark this answers is not in the document]
Comment (rra) [reply src/jwt_authorizer/handlers/tokens.py:155]: Yeah, or if someone posts directly with capabilities that don't exist, or any other random problem that might happen if we make the form more complex.

File: src/jwt_authorizer/handlers/tokens.py [hunk 26]
```diff
@@ -0,0 +1,276 @@
...
+    new_token: Dict[str, object] = {"scope": scope}
+    email = decoded_token.get("email")
+    user = decoded_token.get(config.username_key)
+    uid = decoded_token.get(config.uid_key)
+    if email:
+        new_token["email"] = email
+    if user:
+        new_token[config.username_key] = user
+    if uid:
+        new_token[config.uid_key] = uid
```
[the remark this answers is not in the document]
Comment (rra) [reply src/jwt_authorizer/handlers/tokens.py:172]: This one is in the backward compatibility camp as well, but that's a good point. I've added it to my to-do list for cleaning up when I break backward compatibility.

File: src/jwt_authorizer/handlers/tokens.py [hunk 27]
```diff
@@ -0,0 +1,276 @@
...
+    """
+    config: Config = request.config_dict["jwt_authorizer/config"]
+    logger: Logger = request["safir/logger"]
+
+    try:
+        encoded_token = request.headers["X-Auth-Request-Token"]
+        decoded_token = await authenticate(request, encoded_token)
+    except PyJWTError as e:
+        logger.exception("Failed to authenticate token")
+        raise unauthorized(request, "Invalid token", str(e))
```
[the remark this answers is not in the document]
Comment (rra) [reply src/jwt_authorizer/handlers/tokens.py:149]: Decorator is the plan -- just haven't gotten to it yet, but it's on the list of things to do.

File: src/jwt_authorizer/verify.py (hunk 1 above)
Comment (rra) [reply src/jwt_authorizer/verify.py:91]: Yup, good call.

File: src/jwt_authorizer/verify.py (hunk 3 above)
Comment (rra) [reply src/jwt_authorizer/verify.py:302]: Yeah, it's set in `create_app` if the `key_client` parameter is passed, which is done for the test suite. This is all cleaned up in my next PR.

File: tests/handlers/analyze_test.py (hunk 4 above)
Comment (rra) [reply tests/handlers/analyze_test.py:23]: Yup, good catch!

File: tests/handlers/analyze_test.py (hunk 5 above)
Comment (rra) [reply tests/handlers/analyze_test.py:54]: Ticket, thanks. Fixed the comment.

File: tests/handlers/auth_test.py (hunk 6 above)
Comment (rra) [reply tests/handlers/auth_test.py:54]: Yeah, I want to make sure that failing to have a token at all didn't cause some 500 error inside the code. Python is somewhat vulnerable to that.

File: tests/handlers/auth_test.py (hunk 8 above)
Comment (rra) [reply tests/handlers/auth_test.py:107]: Yes. I definitely want to do that -- my tentative thought is to start reporting all authentication decisions via Kafka so that we can do statistics and easily see what's going on, although it may be easier to do it with structured logging and ELK instead.

File: tests/handlers/auth_test.py (hunk 9 above)
Comment (rra) [reply tests/handlers/auth_test.py:146]: The idea is that either `exec:admin` or `exec:test` is sufficient, so as a test we ask for both with satisfy `any` and then pass in credentials that only have `exec:test` and make sure they're accepted but also don't erroneously claim we have `exec:admin`.

File: tests/handlers/auth_test.py (hunk 10 above)
Comment (rra) [reply tests/handlers/auth_test.py:175]: Yes, indeed. Added a comment.

File: tests/handlers/auth_test.py (hunk 11 above)
Comment (rra) [reply tests/handlers/auth_test.py:213]: Good catch. Added.

File: tests/handlers/auth_test.py (hunk 12 above)
Comment (rra) [reply tests/handlers/auth_test.py:342]: This is a good idea. I'm going to defer this to a later PR because there's a bunch of stuff stacked up, but I will add refactoring the test logic to my todo. I'm not really happy about how magical all these test parameters are to the tests right now.
